error-curation.md

```diff
@@ -0,0 +1,87 @@
+# Error Curation
+
+## Overview
+
+Many errors occur in Temporal yet pithy error messages alone are not enough to give context. We need to provide
+easy-to-reference identifiers to each and pages on each to help.
+
+**Goals**
+
+These are goals for the error curation project, not this proposal in particular though there is overlap.
+
+* Document requirements for rule numbering/identification
+* Document requirements for rule documenting
+* Provide home for all rule documents
+* Build initial set of rules for use in SDKs
+
+**Non-Goals**
+
+* Document every error
+* Make the error documents available on docs.temporal.io at first
+* Provide any UX linking errors from UI, CLI, etc
+* Extremely strict documentation guidelines
+
+## Approach
+
+This approach is inspired by .NET analyzers which have coded values and dedicated pages for each error. For example,
+see:
+
+* https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/categories (e.g.
+  https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1001)
+* https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/DOCUMENTATION.md (e.g.
+  https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1200.md)
+* https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/index.md (e.g.
+  https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD003.md)
+
+Solution:
+
+* A "rule" is referenceable, documented error/warning/note/guideline
+* We are calling these "rules" not "errors"
+  * They may be found eagerly, may just be disable-able warnings, or may even be just best practice rules
+* Rules can be for many things:
+  * Errors (e.g. SDK errors, static analysis errors, server errors, etc)
+  * Guidelines (e.g. best practices, warnings, etc)
+* Every rule has an identifier in the form of `TMPRLXXXX`
+  * "Categories" of rules are given a 100 block as rarely is more needed
+    * Category 100 blocks should try not to be next to another so space can be given in some rare case we exceed 100 in
+      a category
+    * There are no rules on what a category is, just a grouping of like things
+  * The rules within a category do not have to be ordered in any manner. Usually you just use the next available.
+  * `0` is never used in the first two digits
+  * Can discuss the first thousands-block digit though initial thought for first two is:
+    * `TMPRL1XXX` block is for SDK-specific and workflow/activity authoring things
+    * `TMPRL2XXX` block is for server-specific and operation things
+* Rule/error messages across Temporal should be prepended with `"[TMPRLXXXX] "` (sans quotes)
```

Comment:
What about one of these?

```
[TMP1234] Context deadline exceeded. For more info, see: t.mp/1234
[TMP1234] Context deadline exceeded. More info: t.mp/1234
[TMP1234] Context deadline exceeded t.mp/1234
[t.mp/1234] Context deadline exceeded
Context deadline exceeded t.mp/1234
```

My favorite is 4 or 5

Reply:
I think we need to filename-safe number and I wasn't wanting to alter all of our English messages, just a simple prefix

Reply:
If the ID is a URL, then user doesn't need to Google it. Why filename safe? t.mp/1234 could point to a `1234.md` file.

My top goal is having a URL somewhere in error messages. Secondary goal is brevity—of message, code, and URL. If we only have a prefix, then having the URL as the prefix meets both goals.

If we have both a code and a URL, and the URL is at the end of a potentially long message and possibly missed, then it would be useful for the code to be uniquely googleable, and could still keep the URL short:

```
[TMPRL1234] Context deadline exceeded t.mp/1234
```

(I just personally find the longer message and maybe-not-seeing-link-and-googling to be a less good DX)

Reply:
> If the ID is a URL, then user doesn't need to Google it

I'll take the .NET-style/RFC-style/etc-style rule identifier over URL, even if it costs someone a search. It stands out as an error code like others before it when developers glance at the string.

> Why filename safe? t.mp/1234 could point to a 1234.md file.

I have found this easiest when using the existing .NET rules I cited as inspiration. It is easy to reference that string in filename, title, error string, etc. You find this a lot with static analysis tools, which is admittedly where I plan on leveraging this heavily.

> My top goal is having a URL somewhere in error messages.

Wasn't one of my goals listed at the top of the project. Would prefer not to add goals/requirements if we can.

> If we have both a code and a URL

I don't think we need both. But I support easy to reach rules via URL though (and easy to reach via search, and easy to recognize as error codes in strings, etc).

Reply:
>> If the ID is a URL, then user doesn't need to Google it
>
> I'll take the .NET-style/RFC-style/etc-style rule identifier over URL, even if it costs someone a search. It stands out as an error code like others before it when developers glance at the string.

Would all caps stand out enough?

```
[T.MP/1234] Context deadline exceeded
```

>> My top goal is having a URL somewhere in error messages.
>
> Wasn't one of my goals listed at the top of the project. Would prefer not to add goals/requirements if we can.

Since it's so similar and also would involve changing error messages, I think it makes sense to discuss here? Versus eg I open another proposal with the goal of "make it easy for users to get from error message to error page" and we change all the error messages again.

Reply:
When there's a short fix, I like reading it in the message, like:

```
No workflows found in src/workflow.ts. Make sure to export your workflow functions.
```

I'd argue against long explanations since they're harder to read in the place you're reading errors versus a formatted markdown page and more likely to need to be updated, which is easier to do in one place than many.

>> open another proposal with the goal of "make it easy for users to get from error message to error page" and we change all the error messages again.
> 
> I think this proposal is good enough for that use case and we are over-optimizing going any further. 

I think it's good enough for some, but not all, and think it might be worth going further. Would be curious to hear if @mbernier @tlalfano has an opinion or process for answering user questions like this. I don't know the best answer, I think it depends on how many people are in each con category.

### Option A: traditional error codes

```
[TMPRL1234] Context deadline exceeded
```

Con: some won't know to google for TMPRL1234

### Option B: URL error codes

```
[t.mp/1234] Context deadline exceeded
or
[T.MP/1234] Context deadline exceeded
```

Cons: 
- those used to traditional error codes may not see/recognize/pay attention to them
- some may not realize it's a URL

### Option C: traditional error codes + URLs

```
[TMPRL1234] Context deadline exceeded t.mp/1234
```

Con: some may not realize there's a URL, particularly if it's a long error message

### Option D: traditional error codes + explicit URLs

```
[TMPRL1234] Context deadline exceeded. More info: t.mp/1234
```

Cons: 
- longer
- some may not get to the URL if it's a long error message

Pro: `. More info: ` makes it more clear that it's a URL with more info

Reply:
I was afraid of this. There's just too much subjectivity here and it seems like the principled approach just isn't going to win over discussion, voting, options, etc. I was hoping for simple identifiers I could e.g. `#pragma warning disable TMPRL1234` in code and just follow some of that inspiration I linked. But alas it may not be. We have seen this with other projects that just end up taking months instead of the few days they should have (and often end up at a bad answer anyways).

I think it's becoming clear this project isn't ready for a proposal. I will bring it up with team internally and see if we can close this pending more research and/or delegation to another team.

Reply:
I think whether or not you have a URL really depends on the context of what you're doing. I wouldn't want a whole bunch of them showing up in log messages for example. Rust compiler errors just have `error[E0609]: blah` for example, and then more about the location but clippy lints look like this:

```
error: useless conversion to the same type: `temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollActivityTaskQueueResponse`
   --> core/src/worker/activities.rs:632:20
    |
632 |                   Ok(PollActivityTaskQueueResponse {
    |  ____________________^
633 | |                     task_token: vec![2],
634 | |                     activity_id: "act2".to_string(),
635 | |                     ..Default::default()
636 | |                 }
637 | |                 .into())
    | |_______________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
help: consider removing `.into()`
    |
632 ~                 Ok(PollActivityTaskQueueResponse {
633 +                     task_token: vec![2],
634 +                     activity_id: "act2".to_string(),
635 +                     ..Default::default()
636 ~                 })
```

So, I don't think we really need super hard rules here, or maybe we add them over time. Logs / errors / panics maybe just have the code, our static analysis tools maybe behave more like clippy, etc.

I wouldn't let this stop the proposal at all. It's clearly something that can evolve as we implement.

Reply:
I'm very pro this proposal, and would much rather have it implemented now as written than get held up for months in user studies. I don't know the right amount of time to put toward the discussion, versus when to call it and go with what we have. If this is the last pending discussion item and we don't see a quick resolution (internal and/or external slack poll gathering data?), then moving forward with implementation and continuing to discuss this point later and maybe adding a URL at some point SGTM.

Reply:
:+1: FWIW I am not against URL adding in error messages, I just don't want to make it a part of this project and I don't want the rule identifiers to have to be URLs.